src/rmarkdown.ts

```diff
@@ -112,95 +139,275 @@ export class RMarkdownCodeLensProvider implements CodeLensProvider {
       if (isChunkStartLine(lines[line])) {
+        chunkId++;
         chunkStartLine = line;
         chunkLanguage = getChunkLanguage(lines[line]);
+        chunkOptions = getChunkOptions(lines[line]);
+        chunkEval = getChunkEval(chunkOptions);
       }
     } else {
       if (isChunkEndLine(lines[line])) {
-        if (line >= selection.end.line) {
-          if (chunkLanguage === 'r') {
-            const codeRange = new Range(
-              new Position(chunkStartLine + 1, 0),
-              new Position(line - 1, lines[line - 1].length)
-            );
-
-            return runChunksInTerm([codeRange]);
-          }
-        }
+        chunkEndLine = line;
+
+        const chunkRange = new Range(
+          new Position(chunkStartLine, 0),
+          new Position(line, lines[line].length)
+        );
+        const codeRange = new Range(
+          new Position(chunkStartLine + 1, 0),
+          new Position(line - 1, lines[line - 1].length)
+        );
+
+        chunks.push({
+          id: chunkId, // One-based index
+          startLine: chunkStartLine,
+          endLine: chunkEndLine,
+          language: chunkLanguage,
+          options: chunkOptions,
+          eval: chunkEval,
+          chunkRange: chunkRange,
+          codeRange: codeRange
+        });
 
         chunkStartLine = undefined;
+        }
       }
+      line++;
     }
-    line++;
+  return chunks;
+}
+
+function getCurrentChunk(chunks: RMarkdownChunk[], line: number): RMarkdownChunk {
+  const lines = window.activeTextEditor.document.getText().split(/\r?\n/);
+
+  let chunkStartLineAtOrAbove = line;
+  // `- 1` to cover edge case when cursor is at 'chunk end line'
+  let chunkEndLineAbove = line - 1;
+
+  while (chunkStartLineAtOrAbove >= 0 && !isChunkStartLine(lines[chunkStartLineAtOrAbove])) {
+    chunkStartLineAtOrAbove--;
   }
+
+  while (chunkEndLineAbove >= 0 && !isChunkEndLine(lines[chunkEndLineAbove])) {
+    chunkEndLineAbove--;
+  }
+
+  // Case: Cursor is within chunk
+  if (chunkEndLineAbove < chunkStartLineAtOrAbove) {
+    line = chunkStartLineAtOrAbove;
+  } else {
+  // Cases: Cursor is above the first chunk, at the first chunk or outside of chunk. Find the 'chunk start line' of the next chunk below the cursor.
+    let chunkStartLineBelow = line + 1;
+    while (!isChunkStartLine(lines[chunkStartLineBelow])) {
+      chunkStartLineBelow++;
+    }
+    line = chunkStartLineBelow;
+  }
+  const currentChunk = chunks.filter(i => i.startLine === line)[0];
+  return currentChunk;
+}
+
+// Alternative `getCurrentChunk` for cases:
+// - commands (e.g. `goToBeginningOfChunk`) only make sense when cursor is within chunk
+// - when cursor is outside of chunk, no response is triggered for chunk navigation commands (e.g. `goToPreviousChunk`) and chunk running commands (e.g. `runAboveChunks`)
+function getCurrentChunk__CursorWithinChunk(chunks: RMarkdownChunk[], line: number): RMarkdownChunk {
+  let id = 0;
+
+  while (id <= chunks.length - 1) {
```

Comment:
It can use for-of statemente

Reply:
Done in https://github.com/Ikuyadeu/vscode-R/pull/465/commits/213bad2e090c4ae470f019e335b0eedb491c257d